src/Spectre.libDataset.Tests/DatasetTest.cpp

```diff
@@ -0,0 +1,276 @@
+}
+}
+
+// @gmrukwa: provides general tests
+namespace
+{
+/// <summary>
+/// Provides tests for dataset which data source has not yet been destroyed.
+/// </summary>
+/// <seealso cref="::testing::Test" />
+class DatasetInScopeTest : public ::testing::Test
+{
+public:
+    DatasetInScopeTest():
+        data({1,2,3}),
+        entries_metadata({4,5,6}),
+        global_metadata({7}),
+        dataset(data, entries_metadata, global_metadata)
+    { }
+protected:
+    const samples data;
+    const samples_metadata entries_metadata;
+    const dataset_metadata global_metadata;
+    data_set dataset;
+
+    void SetUp() override
+    {
+        dataset = data_set(data, entries_metadata, global_metadata);
+    }
+};
+
+TEST_F(DatasetInScopeTest, provide_the_same_data)
+{
+    EXPECT_EQ(dataset[0], 1);
+    EXPECT_EQ(dataset[1], 2);
+    EXPECT_EQ(dataset[2], 3);
+}
+
+TEST_F(DatasetInScopeTest, provide_the_same_sample_metadata)
+{
+    EXPECT_EQ(dataset.getSampleMetadata(0), 4);
+    EXPECT_EQ(dataset.getSampleMetadata(1), 5);
+    EXPECT_EQ(dataset.getSampleMetadata(2), 6);
+}
+
+TEST_F(DatasetInScopeTest, provide_the_same_metadata)
+{
+    EXPECT_EQ(dataset.getDatasetMetadata()[0], 7);
+}
+
+TEST_F(DatasetInScopeTest, modify_data)
+{
+    ASSERT_NE(dataset[0], 100) << "data was already set to test value";
+    dataset[0] = 100;
+    EXPECT_EQ(dataset[0], 100);
+}
+
+TEST_F(DatasetInScopeTest, modify_sample_metadata)
+{
+    ASSERT_NE(dataset.getSampleMetadata(0), 100) << "metadata was already set to test value";
+    dataset.getSampleMetadata(0) = 100;
+    EXPECT_EQ(dataset.getSampleMetadata(0), 100);
+}
+
+TEST_F(DatasetInScopeTest, modify_metadata)
+{
+    ASSERT_NE(dataset.getDatasetMetadata()[0], 100) << "metadata was already set to test value";
+    dataset.getDatasetMetadata() = { 100 };
+    EXPECT_EQ(dataset.getDatasetMetadata()[0], 100);
+}
+
+TEST_F(DatasetInScopeTest, access_const_sample)
+{
+    const auto& local = dataset;
+    EXPECT_EQ(local[0], dataset[0]);
+}
+
+TEST_F(DatasetInScopeTest, access_const_sample_metadata)
+{
+    const auto& local = dataset;
+    EXPECT_EQ(local.getSampleMetadata(0), dataset.getSampleMetadata(0));
+}
+
+TEST_F(DatasetInScopeTest, modify_data_throws_when_out_of_bounds)
+{
+    EXPECT_THROW(dataset[100] = 100, OutOfRange);
+}
+
+TEST_F(DatasetInScopeTest, modify_sample_metadata_throws_when_out_of_bounds)
+{
+    EXPECT_THROW(dataset.getSampleMetadata(100) = 100, OutOfRange);
+}
+
+TEST_F(DatasetInScopeTest, access_const_sample_throws_when_out_of_bounds)
+{
+    const auto& local = dataset;
+    EXPECT_THROW(local[100], OutOfRange);
+}
+
+TEST_F(DatasetInScopeTest, access_const_sample_metadata_throws_when_out_of_bounds)
+{
+    const auto& local = dataset;
+    EXPECT_THROW(local.getSampleMetadata(100), OutOfRange);
+}
+
+TEST_F(DatasetInScopeTest, access_whole_data_in_readonly)
+{
+    const auto& local = dataset;
+    auto local_data = local.getData();
+    EXPECT_EQ(local_data.size(), 3);
```

Comment:
I'd consider using ASSERT_EQ here, since we risk reading out of bound memory in later commands. (Not really harmful, but just a thing to consider).